CMakeLists.txt

```diff
@@ -9,45 +9,48 @@ cmake_minimum_required(VERSION 2.8)
 # In order to bootstrap the runtime library we need to skip
 # CMake's Fortran tests
 SET(CMAKE_Fortran_COMPILER_WORKS 1)
+SET(CMAKE_Fortran_ABI_COMPILED 0)
+SET(CMAKE_Fortran_COMPILER_SUPPORTS_F90 1)
+SET(CMAKE_Fortran_MODDIR_FLAG "-J ")
```

Comment:
Do you run CMake with `-DCMAKE_Fortran_COMPILER_ID=Flang`? That should set `CMAKE_Fortran_MODDIR_FLAG` correctly.

Reply:
It doesn't for me. (at least on windows)

Reply:
I'm still wondering why these are needed. On Linux, during a clean build (from scratch), `flang` is not yet available or usable, CMake reports the following:
```
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Checking whether /home/b00375952/src/install.release_12x/bin/flang supports Fortran 90
-- Checking whether /home/b00375952/src/install.release_12x/bin/flang supports Fortran 90 - no
```
But it continues on to build the Flang frontend and runtime just fine. What error do you see when building on Windows, if you don't add these lines to `CMakeLists.txt`?

Reply:
```
CMake Error at C:/Users/Administrator/miniconda3/envs/llvm/Library/share/cmake-3.20/Modules/CMakeDetermineCompilerABI.cmake:49 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  C:/Users/Administrator/miniconda3/envs/llvm/Library/share/cmake-3.20/Modules/CMakeTestFortranCompiler.cmake:20 (CMAKE_DETERMINE_COMPILER_ABI)
  runtime/flang/CMakeLists.txt:7 (enable_language)
  ```

Reply:
If second line is not there
```
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_Fortran_PREPROCESS_SOURCE
```

Reply:
If third line is not there, the mod files are not installed to the correct location.

Reply:
I removed the last 2 lines because they are actually set in `runtime/flang/CMakeLists.txt` when `enable_languages(Fortran)` is called.
The first line is still needed.

Reply:
Can you make CMakeError.log available? I believe that the root cause is that `CMAKE_DETERMINE_COMPILER_ABI` cannot find the `flang` executable, which should have been installed when you built and installed classic-flang-llvm-project, and should be reachable via your `PATH` environment variable. The `flang` executable is a symbolic link to `clang` on Linux; I am not sure what form it takes on Windows. If it can be found and executed, CMake should be able to figure out other information about the Flang compiler, and continue without these hacks.

Reply:
```
Compiling the Fortran compiler identification source file "CMakeFortranCompilerId.F" failed.
Compiler: C:/Users/Administrator/miniconda3/envs/llvm/Library/bin/flang.exe 
Build flags: 
Id flags: -v 

The output was:
1
clang version 10.0.1 
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:/Users/Administrator/miniconda3/envs/llvm/Library/bin
 "flang1" CMakeFortranCompilerId.F -opt 0 -terse 1 -inform warn -nohpf -nostatic -cmdline "'+flang -v CMakeFortranCompilerId.F'" -inform warn -ieee 0 -x 19 0x400000 -quad -x 68 0x1 -x 59 4 -x 15 2 -x 49 0x400004 -x 51 0x20 -x 57 0x4c -x 58 0x10000 -x 124 0x1000 -tp px -x 57 0xfb0000 -x 58 0x78031040 -x 47 0x08 -x 48 4608 -x 49 0x100 -def __GCC_ASM_FLAG_OUTPUTS__=1 -def __code_model__=1 -def __amd64__=1 -def __amd64=1 -def __x86_64=1 -def __x86_64__=1 -def __SEG_GS=1 -def __SEG_FS=1 -def __seg_gs=__attribute__((address_space(256))) -def __seg_fs=__attribute__((address_space(257))) -def __REGISTER_PREFIX__= -def __NO_MATH_INLINES=1 -def __SSE2__=1 -def __SSE2_MATH__=1 -def __SSE__=1 -def __SSE_MATH__=1 -def __MMX__=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8=1 -def _WIN32=1 -def _WIN64=1 -def _MSC_VER=1916 -def _MSC_FULL_VER=191627043 -def _MSC_BUILD=1 -def _INTEGRAL_MAX_BITS=64 -def _M_X64=100 -def _M_AMD64=100 -def "__SIZE_TYPE__=long long unsigned int" -def "__PTRDIFF_TYPE__=long long int" -def __LONG_MAX__=2147483647L -def __amd_64__amd64__ -def __k8 -def __k8__ -def __THROW= -def __extension__= -def __PGLLVM__ -preprocess -nofreeform -vect 48 -x 54 1 -x 70 0x40000000 -y 163 0xc0000000 -x 189 0x10 -stbfile CMakeFortranCompilerId-538e6f.stb -modexport CMakeFortranCompilerId-538e6f.cmod -modindex CMakeFortranCompilerId-538e6f.cmdx -output CMakeFortranCompilerId-538e6f.ilm
flang: error: unable to execute command: program not executable
```

Reply:
Odd. On Linux I get these failures too but CMake still manages to make progress:

```
Compiling the Fortran compiler identification source file "CMakeFortranCompilerId.F" failed.
Compiler: /home/b00375952/src/install.release_12x/bin/flang
Build flags:
Id flags: -v

The output was:
1
flang-new version 12.0.0 (https://github.com/flang-compiler/classic-flang-llvm-project d9e7b49409b89afb32142ba67f3f1aa0d8479a98)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/b00375952/src/install.release_12x/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "flang1" CMakeFortranCompilerId.F -opt 0 -terse 1 -inform warn -nohpf -nostatic -cmdline "'+flang -v CMakeFortranCompilerId.F'" -inform warn -ieee 0 -x 19 0x400000 -quad -x 68 0x1 -x 59 4 -x 15 2 -x 49 0x400004 -x 51 0x20 -x 57 0x4c -x 58 0x10000 -x 124 0x1000 -tp px -x 57 0xfb0000 -x 58 0x78031040 -x 47 0x08 -x 48 4608 -x 49 0x100 -stdinc /home/b00375952/src/install.release_12x/bin/../include:/usr/local/include:/home/b00375952/src/install.release_12x/lib/clang/12.0.0/include:/usr/include/x86_64-linux-gnu:/include:/usr/include -def __GCC_ASM_FLAG_OUTPUTS__=1 -def __code_model___=1 -def __amd64__=1 -def __amd64=1 -def __x86_64=1 -def __x86_64__=1 -def __SEG_GS=1 -def __SEG_FS=1 -def __seg_gs=__attribute__((address_space(256))) -def __seg_fs=__attribute__((address_space(257))) -def __REGISTER_PREFIX__= -def __NO_MATH_INLINES=1 -def __SSE2__=1 -def __SSE2_MATH__=1 -def __SSE__=1 -def __SSE_MATH__=1 -def __MMX__=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8=1 -def __SIZEOF_FLOAT128__=16 -def unix=1 -def __unix=1 -def __unix__=1 -def linux=1 -def __linux=1 -def __linux__=1 -def __ELF__=1 -def __gnu_linux__=1 -def __FLOAT128__=1 -def "__SIZE_TYPE__=long unsigned int" -def "__PTRDIFF_TYPE__=long int" -def _LP64=1 -def __LP64__=1 -def __LONG_MAX__=9223372036854775807L -def __amd_64__amd64__ -def __k8 -def __k8__ -def __THROW= -def __extension__= -def __PGLLVM__ -preprocess -nofreeform -vect 48 -x 54 1 -x 70 0x40000000 -y 163 0xc0000000 -x 189 0x10 -stbfile CMakeFortranCompilerId-bce089.stb -modexport CMakeFortranCompilerId-bce089.cmod -modindex CMakeFortranCompilerId-bce089.cmdx -output CMakeFortranCompilerId-bce089.ilm
clang-12: error: unable to execute command: Executable "flang1" doesn't exist!


Detecting Fortran compiler ABI info failed to compile with the following output:
Change Dir: /home/b00375952/src/build.release_12x/flang/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_3b3e4/fast && /usr/bin/make  -f CMakeFiles/cmTC_3b3e4.dir/build.make CMakeFiles/cmTC_3b3e4.dir/build
make[1]: Entering directory '/home/b00375952/src/build.release_12x/flang/CMakeFiles/CMakeTmp'
Building Fortran object CMakeFiles/cmTC_3b3e4.dir/CMakeFortranCompilerABI.F.o
/home/b00375952/src/install.release_12x/bin/flang   -v -c /usr/share/cmake-3.19/Modules/CMakeFortranCompilerABI.F -o CMakeFiles/cmTC_3b3e4.dir/CMakeFortranCompilerABI.F.o
flang-new version 12.0.0 (https://github.com/flang-compiler/classic-flang-llvm-project d9e7b49409b89afb32142ba67f3f1aa0d8479a98)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/b00375952/src/install.release_12x/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "flang1" /usr/share/cmake-3.19/Modules/CMakeFortranCompilerABI.F -opt 0 -terse 1 -inform warn -nohpf -nostatic -cmdline "'+flang -v -c /usr/share/cmake-3.19/Modules/CMakeFortranCompilerABI.F -o CMakeFiles/cmTC_3b3e4.dir/CMakeFortranCompilerABI.F.o'" -inform warn -ieee 0 -x 19 0x400000 -quad -x 68 0x1 -x 59 4 -x 15 2 -x 49 0x400004 -x 51 0x20 -x 57 0x4c -x 58 0x10000 -x 124 0x1000 -tp px -x 57 0xfb0000 -x 58 0x78031040 -x 47 0x08 -x 48 4608 -x 49 0x100 -stdinc /home/b00375952/src/install.release_12x/bin/../include:/usr/local/include:/home/b00375952/src/install.release_12x/lib/clang/12.0.0/include:/usr/include/x86_64-linux-gnu:/include:/usr/include -def __GCC_ASM_FLAG_OUTPUTS__=1 -def __code_model___=1 -def __amd64__=1 -def __amd64=1 -def __x86_64=1 -def __x86_64__=1 -def __SEG_GS=1 -def __SEG_FS=1 -def __seg_gs=__attribute__((address_space(256))) -def __seg_fs=__attribute__((address_space(257))) -def __REGISTER_PREFIX__= -def __NO_MATH_INLINES=1 -def __SSE2__=1 -def __SSE2_MATH__=1 -def __SSE__=1 -def __SSE_MATH__=1 -def __MMX__=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4=1 -def __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8=1 -def __SIZEOF_FLOAT128__=16 -def unix=1 -def __unix=1 -def __unix__=1 -def linux=1 -def __linux=1 -def __linux__=1 -def __ELF__=1 -def __gnu_linux__=1 -def __FLOAT128__=1 -def "__SIZE_TYPE__=long unsigned int" -def "__PTRDIFF_TYPE__=long int" -def _LP64=1 -def __LP64__=1 -def __LONG_MAX__=9223372036854775807L -def __amd_64__amd64__ -def __k8 -def __k8__ -def __THROW= -def __extension__= -def __PGLLVM__ -preprocess -nofreeform -vect 48 -x 54 1 -x 70 0x40000000 -y 163 0xc0000000 -x 189 0x10 -stbfile CMakeFortranCompilerABI-0004c9.stb -modexport CMakeFortranCompilerABI-0004c9.cmod -modindex CMakeFortranCompilerABI-0004c9.cmdx -output CMakeFortranCompilerABI-0004c9.ilm
clang-12: error: unable to execute command: Executable "flang1" doesn't exist!
CMakeFiles/cmTC_3b3e4.dir/build.make:84: recipe for target 'CMakeFiles/cmTC_3b3e4.dir/CMakeFortranCompilerABI.F.o' failed
make[1]: *** [CMakeFiles/cmTC_3b3e4.dir/CMakeFortranCompilerABI.F.o] Error 1
make[1]: Leaving directory '/home/b00375952/src/build.release_12x/flang/CMakeFiles/CMakeTmp'
Makefile:140: recipe for target 'cmTC_3b3e4/fast' failed
make: *** [cmTC_3b3e4/fast] Error 2
```
What happens if you add `-DCMAKE_Fortran_COMPILER_WORKS=ON` when invoking CMake?

Reply:
Doesn't work either

Reply:
OK, I'll rest my case for now. 😄